### PR TITLE
Return a stream of provider responses at the protocol layer

### PR DESCRIPTION
**Pull request by kim**
This provides the online part of `Browse::query` from #82. As stated there, the protocol may also have cached information about providers, which is not part of this patch.

### CODE REVIEW

File: librad/src/net/gossip.rs [hunk 1]
```diff
@@ -67,11 +67,28 @@ pub use types::*;
...
+#[derive(Clone)]
+pub enum Control<A> {
     SendAdhoc { to: PeerInfo, rpc: Rpc<A> },
     Connect { to: PeerInfo, hello: Rpc<A> },
     Disconnect(PeerId),
 }
 
+#[derive(Clone)]
+pub enum Info<A> {
+    Has(Has<A>),
```
[the remark this answers is not in the document]
Comment (kim) [reply librad/src/net/gossip.rs:83]: So I can reuse the `Has` struct and hand it out to the requester

**Comment by kim** [comment]
The indirection for `ProtocolEvent` is simply to make filtering the stream more convenient— we can also create two different streams, but that’s an optimisation for when we need it.

Conflating query with converging the local storage is probably a bad idea, as the storage can’t know _how much_ to fetch. This requires a new gossip message variant, though.

**Comment by FintanH** [comment]
Ok, I think that's all clear.

**Comment by kim** [comment]
Actually, scrap that last sentence: once a peer `Has` a project, the query can be answered from local storage (it can only go away by that peer deleting the project, which we will notice next time we try to fetch from it). Caching and naming things.

**Pull request merged**